Docs/StyleGuide/Style-Guide.md

```diff
@@ -0,0 +1,280 @@
+This style guide provides coding conventions for the Dafny code.
+
+*This documentation is still in progress. Please feel free to add any more suggestions if any.*
+
+## Naming Convention
+Any **variables** are named with `camelCase`.
+```dafny
+var minValue := 1;
+var cipherMessage := "Hello World";
+```
+
+Any **lemmas**, **predicates**, **functions**, **methods**, **classes**, and **modules** are named with `PascalCase`.
+```dafny
+method FindIndex(arr: seq<int>, k: int)
+    ...
+```
+
+Any static or global **constants** are named with `UPPERCASE_WITH_UNDERSCORES`.
+```dafny
+static const MONTHS_IN_A_YEAR := 12;
+```
+
+### Method Prefix
+Avoid redundant names when variables or methods are in a class/module.
+```dafny
+class Integer {
+
+    // The following method converts the given integer
+    // to a string.
+    //
+    // this method name can be simplified to ToString()
+    // so that the method call is Integer.ToString(i)
+    // instead of Integer.IntegerToString(i).
+
+    // YES
+    method ToString(i: int) returns (s: string)
+        ...
+
+    // NO
+    method IntegerToString(i: int) returns (s: string)
+        ...
+}
+```
+
+## Code Layout
+
+### Braces
+Dafny by default opens a new brace in the same line.
+```dafny
+module M() {
+    ...
+    method Met() {
+        ...
+    }
+}
+```
+In case the method (or function, lemma, etc) signature is too long to fit in one line, or in case the signature has at
+least one specification clause, the new brace goes to the new line.
+
+```dafny
+module M() {
+    ...
+    method Met(i: int) returns (j: int)
+        requires i % 2 == 0
+        ensures j > 10
+    {
+        ...
+    }
+}
+```
+
+This applies to every scope: `module`, `class`, `predicate`, `if`, `while`, and more.
+
+### Imports
+Import the modules with `opened` to avoid typing a full module name each time a method is invoked.
+```dafny
+import opened StandardLibrary
+...
+```
+
+When there are more than two modules with the same method name, specify the name of the module during method invocation.
+
+```dafny
+import opened MyModule
+import opened YourModule
+...
+method MyMethod() {
+    MyModule.foo();
+    YourModule.foo();
+}
+```
+
+In case typing the full module name is cumbersome, import it with a shorthanded name.
+
+```dafny
+import M = MyModuleWithACumbersomeName
+import Y = YourModuleWithACumbersomeName
+...
+method MyMethod() {
+    M.foo();
+    Y.foo();
+}
+```
+
+Common imports, such as `StandardLibrary` and `Native`, should be grouped together, followed by custom module imports
+with a blank line in-between.
+
+```dafny
+import opened StandardLibrary
+import opened Native
+
+import opened Donut
+import Coffee
+```
+
+## Indentation and Line Breaks
+
+### Tabs or Spaces?
+Spaces are preferred over tabs. Tabs should only be used to remain consistent with the existing code with tabs.
+
+Use 4 spaces for each indentation.
```

Comment:
Whoa, why?  I thought all decent languages are 2.  I can only think of Java and Python (neither of which qualify) that use 4.

Reply:
Rust uses 4. I was a proponent for 2 for a long time, but would vote for 4 for Dafny.

I guess one of the reasons I prefer 4 is that after a `method` (etc.) there are usually more long keywords like `requires`, `ensures`, or `modifies` on the lines right after without an empty line separating them. 2 spaces doesn't give me as much visual separation as 4.

Reply:
(Thanks for Rust example)

I guess it depends on the code you're writing.  I mainly just want to write plain functional programs, and having 4 space indent will be sad for this.  I'll cede the point, but won't promise to follow it when it doesn't make sense, which seems to be almost always in the code we've been writing.

Reply:
I completely see the point. I can live with either.

Reply:
I like 2. 2 is sufficient for my eyes to clearly notice the indentation, and for deeper levels of indentation, more things stay on my screen with 2.